akka-remote/src/main/scala/akka/remote/artery/tcp/SSLEngineProvider.scala

```diff
@@ -84,19 +85,37 @@ class ConfigSSLEngineProvider(protected val config: Config, protected val log: M
   val SSLRequireMutualAuthentication: Boolean = config.getBoolean("require-mutual-authentication")
   val HostnameVerification: Boolean = config.getBoolean("hostname-verification")
 
-  private lazy val sslContext: SSLContext = {
-    // log hostname verification warning once
-    if (HostnameVerification)
-      log.debug("TLS/SSL hostname verification is enabled.")
-    else
-      log.info(
-        LogMarker.Security,
-        "TLS/SSL hostname verification is disabled. See Akka reference documentation for more information.")
-
-    constructContext()
+  private val SSLContextCacheTime: FiniteDuration = config.getDuration("ssl-context-cache-ttl").toMillis.millis
+
+  // log hostname verification warning once
+  if (HostnameVerification)
+    log.debug("TLS/SSL hostname verification is enabled.")
+  else
+    log.info(
+      LogMarker.Security,
+      "TLS/SSL hostname verification is disabled. See Akka reference documentation for more information.")
+
+  private def sslContext: SSLContext = getFromCache
+
+  private val contextRef = new AtomicReference[Option[CachedSSLContext]](None)
+
+  private case class CachedSSLContext(cached: SSLContext, expires: Deadline)
+
+  private def getFromCache: SSLContext = {
+    contextRef.get() match {
+      case Some(CachedSSLContext(oldContext, expired)) if expired.isOverdue() =>
+        val context = constructContext()
+        contextRef.set(Some(CachedSSLContext(context, SSLContextCacheTime.fromNow)))
+        context
+      case Some(CachedSSLContext(cached, _)) => cached
+      case None =>
+        val context = constructContext()
+        contextRef.set(Some(CachedSSLContext(context, SSLContextCacheTime.fromNow)))
+        context
+    }
   }
 
-  private def constructContext(): SSLContext = {
+  private def constructContext(): SSLContext = synchronized {
```

Comment:
Getting the new keyManagers and trustManagers has to be synchronous in this API.

That means in the use case where keys are rotated, SSLEngineProvider would periodically (based on configuration) rebuild the SSLContext from the 'current' keys, and there would be another process that has to make sure the keys are refreshed in time.

That gets pretty timing-sensitive and arguably error-prone - I think it would be nicer if the (likely asynchronous) process that talks to Vault (or whatever key distribution mechanism is used) can proactively trigger the keys to be rotated when it knows there are new ones. I guess that could be a later addition to this API, though.